autoload/maktaba/json.vim

```diff
@@ -0,0 +1,211 @@
+    let l:value = eval(l:str_match)
+    return [l:value, s:Consume(a:json, len(l:str_match))]
+  endif
+  " First character if any, empty string otherwise.
+  let l:first_char = a:json[0:0]
+  " List
+  if l:first_char is# '['
+    let [l:items, l:remaining] = s:ParseListPartial(a:json, a:custom_values)
+    return [l:items, l:remaining]
+  endif
+  " Dict
+  if l:first_char is# '{'
+    let [l:dict, l:remaining] = s:ParseDictPartial(a:json, a:custom_values)
+    return [l:dict, l:remaining]
+  endif
+  throw maktaba#error#BadValue('Input is not valid JSON text.')
+endfunction
+
+function! s:ParseListPartial(json, custom_values) abort
+  let l:remaining = s:Consume(a:json, 1)
+  if l:remaining[0:0] is# ']'
+    return [[], s:Consume(l:remaining, 1)]
+  endif
+  let l:items = []
+  while !empty(l:remaining)
+    " Parse and consume one value as the next list item.
+    let [l:item, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    call add(l:items, l:item)
+    " If next character is "]", consume and finish list.
+    if l:remaining[0:0] is# ']'
+      let l:remaining = s:Consume(l:remaining, 1)
+      break
+    endif
+    " Only other acceptable character is comma. Consume and continue.
+    if l:remaining[0:0] isnot# ','
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON array item: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+  endwhile
+  return [l:items, l:remaining]
+endfunction
+
+function! s:ParseDictPartial(json, custom_values) abort
+  let l:remaining = s:Consume(a:json, 1)
+  if l:remaining[0:0] is# '}'
+    return [[], s:Consume(l:remaining, 1)]
+  endif
+  let l:dict = {}
+  while !empty(l:remaining)
+    " Parse and consume one value as the key.
+    let [l:key, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    if !maktaba#value#IsString(l:key)
+      throw maktaba#error#BadValue(
+          \ 'Non-string keys not allowed for JSON dicts: %s.', string(l:key))
+    endif
+    " Consume separating ":".
+    if l:remaining[0:0] isnot# ':'
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON key: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+    " Parse and consume one value as the item value.
+    let [l:value, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    let l:dict[l:key] = l:value
+    unlet l:value
+
+    " If next character is "}", consume and finish list.
+    if l:remaining[0:0] is# '}'
+      let l:remaining = s:Consume(l:remaining, 1)
+      break
+    endif
+    " Only other acceptable character is comma. Consume and continue.
+    if l:remaining[0:0] isnot# ','
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON object item: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+  endwhile
+  return [l:dict, l:remaining]
+endfunction
+
+function! s:SetDifference(a, b) abort
```

Comment:
Eventually. I usually defer making it an official util until the 2nd time I need it.